### ISSUE TITLE
--network=auto before vc subcommand

### ISSUE THREAD

**Issue by ajsutton** [Consensys-Incorporated/teku#4471]
### Description
The `auto` option for `--network` is only supported for the validator client subcommand and is expected to become the only supported value in time.  Currently `teku vc --network=auto` works correctly but `teku --network=auto vc` still prints a deprecation warning to use `--network=auto` and defaults to `--network=mainnet`.

Ideally the `--network` option wouldn't have been valid prior to the `vc` subcommand but given it is we want to avoid breaking backwards compatibility.  So likely have to find a way to support `teku --network=auto vc`.

**Comment by ermyas** [comment]
hi @ajsutton - I can take this on if that is OK.

I believe the issue here is that if an option is provided before the subcommand, PicoCli interprets it as an option given to the parent command and not the subcommand. So the `vc` subcommand currently sees the network option as empty and then defaults to using `mainnet`. Because of this, any network value provided before the subcommand is currently ignored e.g. `teku --network=prater vc` would use mainnet instead. 

The easy fix to address the specific case you cited would have been to default to using `auto` whenever the `network` option of the subcommand is empty. But as discussed in #3530 this would break backward compatibility since the current expectation of users is that not specifying a network option defaults to mainnet. 

One option is that we attempt to inherit the parent command's network option if the network option to the vc subcommand is empty. I imagine this is risky however since the user might have expectations that providing options before or after the subcommand is generally equivalent for other cases, and supporting this consistently sounds problematic. Note that this wouldn't work for subcommand specific options.

Perhaps another option to consider is simply that we detect this scenario where a user provides the network option before the subcommand and print a help message about the correct usage of the option and exit.

**Comment by ajsutton** [comment]
That would be great if you can.

I think the first big question is what happened previously if you specified `teku --network=pyrmont vc` - if that still defaulted to `mainnet` then we don't really have a backwards compatibility concern as it never worked correctly.  In which case we could just check in the vc subcommand if the parent command `--network` option has been changed from the default and fail with an error as the user has very likely made an unintentional configuration error.

The second issue that's popped up from this is that specifying `network` in a configuration file used with the validator client subcommand doesn't seem to apply the change to the subcommand but only to the parent command. Again this may have always been a problem which would give us a bit more flexibility.

I actually wonder if the right thing for the `--network` option is to just move to always using `auto` for the validator client. I can't think of any case where users would actually want to use a different network configuration to the beacon node as it won't work correctly.  But potentially this could be an issue with other options as well (eg --data-dir).

So basically I think there's a bunch of research to be done on this to work out what the right answer is.

**Comment by ermyas** [comment]
> I think the first big question is what happened previously if you specified `teku --network=pyrmont vc` - if that still defaulted to `mainnet` then we don't really have a backwards compatibility concern as it never worked correctly.

Yes, I checked, the behaviour was the same prior to the change (#4324) i.e. the network option would be ignored by the vc subcommand if specified before the subcommand.

> The second issue that's popped up from this is that specifying `network` in a configuration file used with the validator client subcommand doesn't seem to apply the change to the subcommand but only to the parent command. Again this may have always been a problem which would give us a bit more flexibility.

I might be missing something on this one. Providing the network in a config file for the vc sub-command seems to work fine for me (whether specified before or after the subcommand and when using `auto` as an option value or something else). 

> I actually wonder if the right thing for the `--network` option is to just move to always using `auto` for the validator client.

Right, I believe this was already in the gradual 3-step transition plan you outlined for this [here](https://github.com/ConsenSys/teku/issues/3530#issuecomment-913242211)? We opted not to make "auto" the default right away because of backward compatibility reasons but planned to eventually do so as a second step and then finally remove the `network` option for `vc` altogether. As you noted, perhaps a scenario in which a user might want to specify different network configurations than the beacon node's probably doesn't make sense, so accommodating this for the sake of backward compatibility might not be worthwhile. I will look into the potential impact of this.

**Comment by ajsutton** [comment]
> Yes, I checked, the behaviour was the same prior to the change (#4324) i.e. the network option would be ignored by the vc subcommand if specified before the subcommand.

Ok that's good that we didn't unexpectedly make things worse.

> I might be missing something on this one. Providing the network in a config file for the vc sub-command seems to work fine for me (whether specified before or after the subcommand and when using `auto` as an option value or something else).

Hmm, interesting. The user that reported this on discord said they were seeing this problem when using a config file - I haven't actually double checked that myself.  We could probably write a unit test by extending `AbstractBeaconNodeCommandTest` to prove how the different scenarios work in terms of the resulting config.

> Right, I believe this was already in the gradual 3-step transition plan you outlined for this [here](https://github.com/ConsenSys/teku/issues/3530#issuecomment-913242211)? We opted not to make "auto" the default right away because of backward compatibility reasons but planned to eventually do so as a second step and then finally remove the `network` option for `vc` altogether. As you noted, perhaps a scenario in which a user might want to specify different network configurations than the beacon node's probably doesn't make sense, so accommodating this for the sake of backward compatibility might not be worthwhile. I will look into the potential impact of this.

Yep exactly.  I think taking the next step now of defaulting to `auto` makes sense.  It will be a fair while before we can remove the `--network` option though as I think teku then won't start because of the unrecognised command (if it's after the `vc`).

**Comment by ermyas** [comment]
Apologies for the delay on this @ajsutton. I got tied up with a few things. I've submitted PR #4629
